examples/lb3-application/lb3app/test/acceptance.js

```diff
@@ -0,0 +1,132 @@
+require('should');
+
+let app;
+
+function json(verb, url) {
+  // use the original app's server
+  return request(app.server)
+    [verb](url)
+    .set('Content-Type', 'application/json')
+    .set('Accept', 'application/json')
+    .expect('Content-Type', /json/);
+}
+
+describe('LoopBack 3 style acceptance tests', function () {
+  before(async function () {
+    app = new ExpressServer();
+    await app.boot();
+    await app.start();
+  });
+
+  after(async () => {
+    await app.stop();
+  });
+
+  context('basic REST calls for LoopBack 3 application', () => {
+    it('creates and finds a CoffeeShop', function (done) {
+      json('post', '/api/CoffeeShops')
+        .send({
+          name: 'Coffee Shop',
+          city: 'Toronto',
+        })
+        .expect(200)
+        .end(function (err, res) {
+          assert(typeof res.body === 'object');
+          assert(res.body.name);
+          assert(res.body.city);
+          assert.equal(res.body.name, 'Coffee Shop');
+          assert.equal(res.body.city, 'Toronto');
+          done();
+        });
+    });
+
+    it("gets the CoffeeShop's status", function (done) {
+      json('get', '/api/CoffeeShops/status').expect(200, function (err, res) {
+        res.body.status.should.be.equalOneOf(
+          'We are open for business.',
+          'Sorry, we are closed. Open daily from 6am to 8pm.',
+        );
+        done();
+      });
+    });
+
+    it('gets external route in application', function (done) {
+      json('get', '/ping').expect(200, function (err, res) {
+        assert.equal(res.text, 'pong');
+        done();
+      });
+    });
+  });
+
+  context('authentication', () => {
+    let User;
+
+    before(() => {
+      User = lb3App.models.User;
+    });
+
+    it('creates a User and logs them in and out', function (done) {
+      User.create({email: 'new@email.com', password: 'L00pBack!'}, function (
+        err,
+        user,
+      ) {
+        assert.equal(user.email, 'new@email.com');
+        User.login(
+          {
+            email: 'new@email.com',
+            password: 'L00pBack!',
+          },
+          function (err2, token) {
+            token.should.have.properties('ttl', 'userId', 'created', 'id');
+            assert.equal(token.userId, user.id);
+            User.logout(token.id);
+            User.deleteById(user.id);
+          },
+        );
+        done();
+      });
+    });
+
+    it('rejects anonymous requests to protected endpoints', function (done) {
+      json('get', '/api/CoffeeShops/greet').expect(401, function (err, res) {
+        assert.equal(res.body.error.code, 'AUTHORIZATION_REQUIRED');
+      });
+      done();
+    });
+
+    it('makes an authenticated request', function (done) {
+      User.create({email: 'new@email.com', password: 'L00pBack!'}, function (
+        err,
+        user,
+      ) {
+        user.email.should.be.equal('new@email.com');
+        json('post', '/api/users/login')
```

Comment:
Please find a way how to allow this test to call `User.login` directly via JavaScript API, so that users don't have to rework all their tests calling model API from JavaScript to make HTTP requests instead.

Reply:
Is this perhaps trying to workaround the problem that `User` has `unknown` type? (See the discussion below.)

Reply:
IIUC...I think Nora tests the endpoint instead of the API here because it's acceptance tests. The APIs are tested in the integration test file.

Reply:
I realize now this file is a bit confusing because I was trying to mimic [the original acceptance tests](https://github.com/strongloop/loopback-next/blob/master/examples/lb3-application/src/__tests__/acceptance/lb3app.acceptance.ts) which makes this now acceptance + integration but I think I'll put the auth tests in their own file.

Reply:
Here is my point of view: this acceptance test has two parts. In the first part, we want to setup data (e.g. User instance, a valid AccessToken, etc.). In the second part, we test the actual scenario.

I am arguing that for test setup, we should use the most performant solution. This part is not testing anything, it's just preparing data to enable the scenario being tested. That's why I prefer to call model methods directly via JS API. HTTP calls introduce non-negligible slow down.

Reply:
> I am arguing that for test setup, we should use the most performant solution. This part is not testing anything, it's just preparing data to enable the scenario being tested. That's why I prefer to call model methods directly via JS API. HTTP calls introduce non-negligible slow down.

Oh I see, thanks for explanation 👍